src/sentry/tasks/releasemonitor.py

```diff
@@ -0,0 +1,178 @@
+                    )
+                    .set_limit(CHUNK_SIZE + 1)
+                    .set_offset(offset)
+                    .set_totals(True)
+                )
+                results = snuba.raw_snql_query(
+                    query, referrer="tasks.process_projects_with_sessions"
+                )
+                data = results["data"]
+                count = len(data)
+                more_results = count >= CHUNK_SIZE
+                offset += count
+                total_sessions = results["totals"]["sessions"]
+
+            # 3. Using the sums from #2, calculate adoption rate (relevant sessions / all sessions) update the appropriate ReleaseProjectEnvironment model adopted/unadopted fields.
+            with metrics.timer(
+                "sentry.tasks.monitor_release_adoption.process_projects_with_sessions.updates"
+            ):
+                for row in data:
+                    try:
+                        if row["sessions"] / total_sessions >= REQUIRED_ADOPTION_PERCENT:
+                            rpe = ReleaseProjectEnvironment.objects.get(
+                                project_id=row["project_id"],
+                                release_id=Subquery(
+                                    Release.objects.filter(
+                                        organization=org_id, version=row["release"]
+                                    ).values("id")[:1],
+                                ),
+                                environment__name=row["environment"],
+                                environment__organization_id=org_id,
+                            )
+                            adopted_ids.append(rpe.id)
+                            if rpe.adopted is None:
+                                rpe.update(adopted=timezone.now())
+                    except ReleaseProjectEnvironment.DoesNotExist:
+                        logger.info(
+                            "process_projects_with_sessions.skipped_update",
+                            extra={
+                                "project_id": row["project_id"],
+                                "environment": row["environment"],
+                                "release": row["release"],
+                                "reason": "ReleaseProjectEnvironment.DoesNotExist",
+                            },
+                        )
+
+            if not more_results:
+                break
+        else:
+            logger.info(
+                "process_projects_with_sessions.loop_timeout",
+                extra={"org_id": org_id, "project_ids": project_ids},
+            )
+
+    # 4. Cleanup - releases that are marked as adopted that didn’t get any results in #2 need to be marked as unadopted
+    ReleaseProjectEnvironment.objects.filter(
+        project_id__in=project_ids, unadopted__isnull=True
+    ).exclude(Q(adopted=None) | Q(id__in=adopted_ids)).update(unadopted=timezone.now())
```

Comment:
I forget which indexes we added here, but we might end up needing one like `(project_id, adopted, unadopted)`. We can handle this in a future PR, but we should run this query on some larger orgs as part of testing and make sure this isn't really slow - for orgs with 100ks of releases we might need it.